src/shims/unix/linux/fd/event.rs

```diff
@@ -35,4 +38,32 @@ impl FileDescriptor for Event {
     ) -> InterpResult<'tcx, io::Result<i32>> {
         Ok(Ok(0))
     }
+
+    /// A write call adds the 8-byte integer value supplied in
+    /// its buffer to the counter.  The maximum value that may be
+    /// stored in the counter is the largest unsigned 64-bit value
+    /// minus 1 (i.e., 0xfffffffffffffffe).  If the addition would
+    /// cause the counter's value to exceed the maximum, then the
+    /// write either blocks until a read is performed on the
+    /// file descriptor, or fails with the error EAGAIN if the
+    /// file descriptor has been made nonblocking.
+
+    /// A write fails with the error EINVAL if the size of the
+    /// supplied buffer is less than 8 bytes, or if an attempt is
+    /// made to write the value 0xffffffffffffffff.
+    ///
+    /// FIXME: use endianness
+    fn write<'tcx>(
+        &self,
```

Comment:
Is it possible to change `write` to take `&mut self` and then avoid the `Cell`?

Reply:
Nope, we looked into this, and the issue is that preexisting code needs to write access memory while using the file descriptors. It would be possible, but it would make all code using writes very roundabout. Also I think other write impls already use interior mutability.

I can open a PR if you'd like to see the workarounds needed

Reply:
So... one write impl that "accidentally" uses interior mutability is our actual `File` handles. `Write` is implemented for `&File`.

The reason we can't easily switch to `&mut` is this piece of code:

https://github.com/rust-lang/miri/blob/master/src/shims/unix/fs.rs#L814-L819

We could handle all of this correctly, but it would require cloning somewhere, as we can't hold the mutable reference to the file descriptor while also reading from memory.

Reply:
Yeah we're doing something rather clever there, directly copying from the machine memory to the syscall.